src/compiler/core.ts

```diff
@@ -1787,14 +1787,21 @@ namespace ts {
      *        (0.4 allows 1 substitution/transposition for every 5 characters,
      *         and 1 insertion/deletion at 3 characters)
      */
-    export function getSpellingSuggestion<T>(name: string, candidates: T[], getName: (candidate: T) => string | undefined): T | undefined {
+    export function getSpellingSuggestion<T>(name: string, candidates: T[], getName: (candidate: T) => string | undefined, isJsxAttr: boolean): T | undefined {
         const maximumLengthDifference = Math.min(2, Math.floor(name.length * 0.34));
         let bestDistance = Math.floor(name.length * 0.4) + 1; // If the best result isn't better than this, don't bother.
         let bestCandidate: T | undefined;
         let justCheckExactMatches = false;
         const nameLowerCase = name.toLowerCase();
         for (const candidate of candidates) {
             const candidateName = getName(candidate);
+            if (isJsxAttr) {
+                const htmlFor = name === "for" && candidateName === "htmlFor";
+                const className = name === "class" && candidateName === "className";
+                if (htmlFor || className) {
+                    return candidate;
+                }
+            }

```

Comment:
Also another question then this raises is that should we change the algorithm to prefer the name that starts or ends with the name we are looking for ?

Reply:
Is this question for me to resolve? 🙈

Reply:
This is the question for the team.. including @weswigham @sandersn @RyanCavanaugh @DanielRosenwasser to comment on this.

Reply:
You can essentially get this by reweighting the distance heuristic - @sandersn had strong opinions last time we talked about it being not usually meaningfully different, though. Perhaps prefix/suffix matching is a worthwhile pre-distance-calculation selection stage. I do dislike these hardcoded suggestions, and would rather see them pulled out from the candidate list by some well-reasoned rules.

Reply:
+1 on not having to deal with hardcoding things..

Reply:
Can you give me an example? Are you asking whether we want to try to auto-correct something like `classy`?

Reply:
But these two cases are really the special knowledge of JSX.
In the class and for case, we exactly know what developer want is className and htmlFor.
What if the new algorithms cannot offer these two candidates? Or have to overfit the algorithm to provide these two just make them not hardcoded? That'll be somehow strange...

Reply:
I tend to agree that these are truly special cases—since the relationship between `for`/`htmlFor` and `class`/`className` comes from the mapping between HTML attributes and DOM element property names, the spelling similarity is not a coincidence, but it’s not a given either. You can imagine a world where the DOM spec decided to map the `for` HTML attribute onto `HTMLLabelElement['targetInputId']`, or the `class` HTML attribute onto `Element['cascadingStyleSheetClassName']`. Both of these would be terrible APIs, but my point is that despite the fairly small string distance between these substitutions, the string distance is not the reason they should be offered. So I’m unconvinced that they merit a change to the string distance algorithm.

Reply:
I don’t know if I agree .. the current reason for should suggest htmlFor is because it’s similar and looks like something user is wrote wrong .. if the new combination arises where suggestion is different or varies too much then those cases need to be handled separately and before any of this algorithm in that case..

Reply:
It's not a spelling mistake, it is another type of mistake. It should be extracted out but I think it's somehow over engineering when there is only two special cases. it is always possible to refactor this in to a new mistake fixer when it have more cases of non-spelling mistakes.

Reply:
> It's not a spelling mistake, it is another type of mistake.

I agree, but I think it's fine to have a JSX-only suggestion function. We have plenty of JSX-specific cases already which wrap existing functions. See my other comment.